Source/Search/Mixins/AdvancedSearchProviderMixin.lua

```diff
@@ -0,0 +1,225 @@
+  Auctionator.Debug.Message("AuctionatorAdvancedSearchProviderMixin:GetSearchProvider()")
+
+  --Run the query, and save extra filter data for processing
+  return function(searchTerm)
+    C_AuctionHouse.SendBrowseQuery(searchTerm.query)
+    self.currentFilter = searchTerm.extraFilters
+    self.itemKeyInfoQueue = {}
+  end
+end
+
+function AuctionatorAdvancedSearchProviderMixin:HasCompleteTermResults()
+  Auctionator.Debug.Message("AuctionatorAdvancedSearchProviderMixin:HasCompleteTermResults()")
+
+  --Loaded all the terms from API, and we have filtered every item
+  return C_AuctionHouse.HasFullBrowseResults() and #(self.itemKeyInfoQueue) == 0
+end
+
+function AuctionatorAdvancedSearchProviderMixin:OnSearchEventReceived(eventName, ...)
+  Auctionator.Debug.Message("AuctionatorAdvancedSearchProviderMixin:OnSearchEventReceived()", eventName, ...)
+
+  if eventName == "AUCTION_HOUSE_BROWSE_RESULTS_UPDATED" then
+    self:ProcessSearchResults(C_AuctionHouse.GetBrowseResults())
+  elseif eventName == "AUCTION_HOUSE_BROWSE_RESULTS_ADDED" then
+    self:ProcessSearchResults(...)
+  elseif eventName == "ITEM_KEY_ITEM_INFO_RECEIVED" then
+    self:ProcessItemKeyInfo(...)
+  elseif eventName == "AUCTION_HOUSE_BROWSE_FAILURE" then
+    AuctionHouseFrame.BrowseResultsFrame.ItemList:SetCustomError(
+      RED_FONT_COLOR:WrapTextInColorCode(ERR_AUCTION_DATABASE_ERROR)
+    )
+  end
+end
+
+function AuctionatorAdvancedSearchProviderMixin:ProcessSearchResults(addedResults)
+  Auctionator.Debug.Message("AuctionatorAdvancedSearchProviderMixin:ProcessSearchResults()")
+
+  local results = {}
+
+  for index = 1, #addedResults do
+    -- Run filter checks on every item key. Some might not be added to the
+    -- results yet, but when the relevant information arrives in an event
+    if self:FilterByItemLevel(addedResults[index].itemKey) then
+      if self:FilterByExact(addedResults[index].itemKey) then
+        table.insert(results, addedResults[index].itemKey)
+      end
+    end
+  end
+
+  self:AddResults(results)
+end
+
+function AuctionatorAdvancedSearchProviderMixin:ProcessItemKeyInfo(itemID)
+  --Event for missing info received about itemID.
+  for index, itemKey in ipairs(self.itemKeyInfoQueue) do
+    local itemKeyInfo = C_AuctionHouse.GetItemKeyInfo(itemKey)
+    if itemKeyInfo then
+      --Remove key from list of those with missing info
+      table.remove(self.itemKeyInfoQueue, index)
+
+    --Only exact search uses this info, and the event won't have been queued
+    --otherwise.
+      if self:ExactMatchCheck(itemKeyInfo) then
+        self:AddResults({itemKey})
+      else
+        --Post empty results, so the mixin supplying it runs
+        --self:HasCompleteTermResults() and can see if the search is complete
+        self:AddResults({})
+      end
+
+      --Only one new result per event
+      break
+    end
+  end
+end
+
+function AuctionatorAdvancedSearchProviderMixin:FilterByItemLevel(itemKey)
```

Comment:
There's a lot going on in this function, and its hard to understand it given all of the nested boolean expressions. Consider refactoring into something that is a little more intention revealing:
```lua
return ItemLevelPresent(itemKey) and ItemSatisfiesLevelFilters(itemKey))

function AuctionatorAdvancedSearchProviderMixin:ItemSatisfiesLevelFilters(itemKey)
  return
    ( 
      -- Min item level filter satisfied
      self.currentFilter.minItemLevel ~= nil and
      self.currentFilter.minItemLevel < itemKey.itemLevel
    ) or (
      -- Max item level filter satisfied
      self.currentFilter.maxItemLevel ~= nil and
      self.currentFilter.maxItemLevel > itemKey.itemLevel
    ) 
end
```
I may have got the logic wrong, but all the more reason to make this more explicit so that we can understand it later if it needs updating.

Reply:
Yeah, the filter logic is a bit awkward